### PR TITLE
File Block: Add support for embedding PDFs

### LINKED ISSUE #19077: Add block: Document/PDF with Preview
**Is your feature request related to a problem? Please describe.**
CF https://github.com/WordPress/gutenberg/issues/15097

PDF thumbnails (preview images) were introduced in WordPress 4.7. Displaying PDF thumbnails works exactly the same as displaying image thumbnails, via `wp_get_attachment_image()`, as long as `$icon` is set to `false`. This is how the attachment single page displays PDFs and (I assume) how the Media Library itself displays the PDF thumbnails.

These thumbnail images don't show up for the image block or for galleries because the attachment itself isn't an image. If you know the direct URL to the thumbnails (which can be exposed using something like the Display All Image Sizes plugin), you can insert them via link and connect them to the PDF, but this is a clumsy process.

If my clients are any example, many users of WordPress produce a LOT of PDF files, and they want to display them with something more than a link. There's a reason that the [Document Gallery plugin](https://wordpress.org/plugins/document-gallery/) still has 20,000 active installations even though it hasn't been updated since WordPress 4.9 and the [PDF Embedder plugin](https://wordpress.org/plugins/pdf-embedder/) has 200,000 active installations. 

**Describe the solution you'd like**
I want a PDF preview block that automatically inserts the chosen thumbnail size and links it to the PDF file--or even a document block that would insert an icon for a Word doc or PowerPoint or text file. The Media Library does this automatically, so the functionality is obviously in core.
Screenshot_2019-12-11 Media Library ‹ PDF Block Test Site — WordPress

A gallery block for documents would be a nice bonus.

**Describe alternatives you've considered**
I have created a custom block via ACF to use on client sites for now. It has a file upload field and a selector for the image size. Unfortunately, I don't have the JS skills to convert that into a real block and submit it to the repository as a block plugin. But given the number of PDF files I have to deal with, I didn't want to spend a lot of time inserting the thumbnails via URL and linking them ditto.

### PR DESCRIPTION
**Pull request by pento**
## Description
Fixes #19077.

This PR adds an option to the File block, when a PDF file is inserted, to show an embedded version of the PDF.

PDFs are a bit of a special case amongst files: they're the only document file type with native embed support in most major browsers ([mobile devices excepted](https://caniuse.com/#feat=pdf-viewer)).

## How has this been tested?

- Insert a file block and select a PDF from the media library.
- Drag/drop a PDF to upload it.
- Switch from a PDF, to a different file type, and back again.
- Tested in different browsers: Chrome, Firefox, Safari, Edge, IE11.
- Tested on different OSes: MacOS, Windows 10, Android, iOS on iPhone, iOS on iPad.
- Tested on IE11 with and without Acrobat Reader installed.

## Types of changes

New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.
- [x] I've included developer documentation if appropriate.
- [x] I've updated all React Native files affected by any refactorings/renamings in this PRx

### CODE REVIEW

**Comment by pento** [comment]
I've been investigating what can be done for browsers that don't support embedded PDFs. There's no good option here.

Broadly speaking, browsers running on mobile devices are the ones that have a problem.

- Chrome on Android renders a big black block with an "Open" button in the middle, for downloading and viewing the PDF in the PDF reader app.
- Firefox on Android renders a blank space, but there is an extension which can be installed to add embedded PDF support.
- Everything on iOS renders the first page of the PDF, stretched to fit the size of the container element.

Unfortunately, there's no such thing as reliable feature detection for embedded PDF support, which gives us a few options:

- Leave it as is, the major mobile browsers will just have suboptimal (but still usable) behaviour.
- Use [PDF.js](https://mozilla.github.io/pdf.js/) to render PDFs, which will work everywhere, but it's a big download to be adding to all page loads.
- Use user agent detection to determine if the browser probably doesn't support inline PDFs.

1c40e28 implements the third option, but I'm interested in arguments either way. 🙂

**Comment by pento** [comment]
The help text is a little bit wordy, I'd appreciate some thoughts on making it less so.

**Comment by michelleweber** [comment]
Do one of these work for you? Depending on what terms you prefer:

_Note: Most mobile devices do not display embedded PDFs._

or

_Most phone and tablet browsers won't display embedded PDFs._

**Comment by pento** [comment]
Perfect, thanks for your help, @michelleweber!

**Comment by aaronrobertshaw** [comment]
This tests well for me, especially in Chrome, Firefox, Safari and Edge on MacOS. 

IE11 on a Windows 10 VM with the Adobe Acrobat Reader and plugin installed would try to embed the PDF but appear to get stuck on the "initializing" display of the embedded object. Testing via Browserstack using Windows 10 and IE11 everything worked as expected. So it might have been something funky with my VM.

The various mobile and tablet browsers I could test, both on iOS and Android, worked as described.

File: packages/block-library/src/file/index.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * @param array $attributes The block attributes.
+ * @param array $content    The block content.
+ *
+ * @return string Returns the modified block content.
+ */
+function render_block_core_file( $attributes, $content ) {
+	$script = '';
+	if ( ! empty( $attributes['showInlineEmbed'] ) ) {
+		$script = <<<HTML
```
Comment (swissspidy) [inline packages/block-library/src/file/index.php:19]: Why is this script not an external script that is set as `WP_Block_Type::$script` when registering the block  using `register_block_type`? Or alternatively, why not make it an external script and call `wp_enqueue_script` if `! empty( $attributes['showInlineEmbed'] )`?

File: packages/block-library/src/file/save.js [hunk 2]
```diff
@@ -11,11 +12,31 @@ export default function save( { attributes } ) {
...
+						<object
+							className="wp-block-file__embed"
+							data={ href }
+							type="application/pdf"
+							style={ {
+								width: '100%',
+								height: `${ embedHeight }px`,
+							} }
+							aria-label={ sprintf(
+								/* translators: Placeholder is the filename. */
```
Comment (swissspidy) [inline packages/block-library/src/file/save.js:33]:
```suggestion
								/* translators: %s: filename. */
```

Core usually uses this style for translator comments

File: packages/block-library/src/file/index.php (hunk 1 above)
Comment (swissspidy) [reply packages/block-library/src/file/index.php:19]: Also, as an external script it can be de-duped with `packages/block-library/src/file/utils.js`

Comment (pento) [reply packages/block-library/src/file/index.php:19]: `wp_enqueue_script()` is probably the best option for de-duplicating the code, but I'm not aware of that method being used for any other blocks. (Or even the `script` property being used, for that matter.)

It seems like it would need a special case to keep it unbundled, and possibly some additional magic when it's merged to Core?

Comment (pento) [reply packages/block-library/src/file/index.php]: 2b239f3102 is a brief exploration of this: I don't mind it for this PR, I'm not wild about it as a generic solution.

**Comment by mapk** [comment]
I ran through some quick tests.

### Embed reacts a little funky with alignments

alignments

### Viewer menu is lost at minimum height

pdf-viewer

### Embed is lost when grouping the block

grouping

**Comment by pento** [comment]
Thanks for the testing, @mapk!

> Embed reacts a little funky with alignments
> Embed is lost when grouping the block

These both had the same root cause, fixed in 5739c88. 

> Viewer menu is lost at minimum height

This appears to be a bug in Chrome's default PDF viewer. They're working on an upgrade to the PDF viewer (you can see it by installing Chrome Canary and enabling this flag: `chrome://flags/#pdf-viewer-update`) which includes a fix for this.

**Comment by mapk** [comment]
Thanks @pento! Just to clarify, we don't have any control over the browser-based embed UI, correct? The colors (dark grey) and icons are all beyond our control.

If that's the case, this looks good to me. Adding an embed option to this block is a cool feature that I imagine will be useful for many. :)

**Comment by pento** [comment]
> Just to clarify, we don't have any control over the browser-based embed UI, correct? The colors (dark grey) and icons are all beyond our control.

That's correct, the embed UI is determined entirely by the browser (or in the case of Windows/IE, Adobe Acrobat 🙃).

File: packages/block-library/src/file/index.php (hunk 1 above)
Comment (youknowriad) [reply packages/block-library/src/file/index.php]: I see `render_callback` as a function that the sole purpose is to render (return) the content of the block and not trigger a side effect elsewhere. Using `wp_enqueue_script` creates an indirection that is very hard to follow.

I think a lot of people ask ways for providing frontend scripts for blocks and it could be a property in `register_block_type` (the equivalent to `editor_script` => `frontend_script`).

For the purpose of this PR specifically, I don't mind the hack, whether it's inline or `wp_enqueue_script`, both are not ideal solution for me but ok.

File: packages/block-library/src/file/block.json [hunk 3]
```diff
@@ -33,10 +33,17 @@
 			"type": "string",
 			"source": "html",
 			"selector": "a[download]"
+		},
+		"showInlineEmbed": {
```
Comment (youknowriad) [inline packages/block-library/src/file/block.json:37]: Should we call it "embed" or "preview"? I mean it's not that important but we could the same for like images, text .... any file that can be previewed. Something like "quick view" in MacOS

Comment (swissspidy) [reply packages/block-library/src/file/block.json:37]: `showPreview` or `displayPreview` perhaps?

File: packages/block-library/src/file/block.json [hunk 4]
```diff
@@ -33,10 +33,17 @@
...
+		"embedHeight": {
+			"type": "number"
 		}
 	},
 	"supports": {
 		"anchor": true,
 		"align": true
-	}
+	},
+	"script": "utils.js"
```
Comment (youknowriad) [inline packages/block-library/src/file/block.json:48]: What is doing?

Comment (youknowriad) [reply packages/block-library/src/file/block.json:48]: It seems it's a way to tell pack to generate a separate script? Why not just a package because that's our way of doing this in Gutenberg? 

Alternatively, if we implement the `frontend_script` proposal, we could introduce a convention like this (I see it more as a file in the block  folder with the property here though). Having the property here will persist in the block registration... which I think is confusing.

**Comment by paaljoachim** [comment]
Hey Gary @pento 
How is this PR coming along?

**Comment by johngodley** [comment]
Noting that I've seen Safari 14 on Big Sur have issues with displaying a PDF via `object` and it sometimes freezes the page. The issue is caused by the experimental browser feature 'incremental pdf loading', which seems to be enabled by default.

I don't know if it also applies to this PR, but thought it worth mentioning for testing.

**Comment by pento** [comment]
Since #30341 landed, the webpack changes are no longer needed. It was easier to just create a new branch, rather than trying to rebase this one, so I'm closing this PR in favour of #30857.
